pkg/meta/sql.go

```diff
@@ -3677,205 +3677,93 @@ func isDuplicateEntryErr(err error) bool {
+			n.Nlink = 1
 		}
-	} else {
-		cloneEno = m.cloneEntry(ctx, srcIno, srcAttr.Typ, dstParentIno, dstName, &dstIno, cmode, cumask, count, total, true, concurrent)
-	}
-	return cloneEno
-}
 
-func (m *dbMeta) cloneEntry(ctx Context, srcIno Ino, srcType uint8, dstParentIno Ino, dstName string, dstIno *Ino, cmode uint8, cumask uint16, count, total *uint64, attach bool, concurrent chan struct{}) syscall.Errno {
-	var err error
-	switch srcType {
-	case TypeDirectory:
-		srcNode := node{Inode: srcIno}
-		if err = m.txn(func(s *xorm.Session) error {
-			ok, err := s.Get(&srcNode)
-			if err != nil {
+		if top {
+			var pattr Attr
+			var pn = node{Inode: parent}
+			if exist, err := s.Get(&pn); err != nil {
 				return err
-			}
-			if !ok {
+			} else if !exist {
 				return syscall.ENOENT
 			}
-			return m.mkNodeWithAttr(ctx, s, srcIno, &srcNode, dstParentIno, dstName, dstIno, cmode, cumask, attach)
-		}); err != nil {
-			return errno(err)
-		}
-		if err := m.txn(func(s *xorm.Session) error {
-			return mustInsert(s, &dirStats{Inode: *dstIno})
-		}); err != nil {
-			return errno(err)
-		}
-		var entries []*Entry
-		if eno := m.doReaddir(ctx, srcIno, 0, &entries, -1); eno != 0 {
-			return eno
-		}
-		atomic.AddUint64(total, uint64(len(entries)))
-		// try directories first to increase parallel
-		var dirs int
-		for i, e := range entries {
-			if e.Attr.Typ == TypeDirectory {
-				entries[dirs], entries[i] = entries[i], entries[dirs]
-				dirs++
-			}
-		}
-
-		var wg sync.WaitGroup
-		var countDir uint32
-		var errCh = make(chan syscall.Errno, cap(concurrent))
-		for i, entry := range entries {
-			select {
-			case eno := <-errCh:
-				errCh <- eno
-				return eno
-			default:
-			}
-			if entry.Attr.Typ == TypeDirectory {
-				select {
-				case concurrent <- struct{}{}:
-					wg.Add(1)
-					go func(srcIno Ino, dstName string) {
-						defer wg.Done()
-						var dstIno2 Ino
-						eno := m.cloneEntry(ctx, srcIno, TypeDirectory, *dstIno, dstName, &dstIno2, cmode, cumask, count, total, true, concurrent)
-						if eno == 0 {
-							atomic.AddUint32(&countDir, 1)
-						}
-						if eno != 0 && eno != syscall.ENOENT {
-							errCh <- eno
-						}
-						<-concurrent
-					}(entry.Inode, string(entry.Name))
-				default:
-					var dstIno2 Ino
-					if eno := m.cloneEntry(ctx, entry.Inode, TypeDirectory, *dstIno, string(entry.Name), &dstIno2, cmode, cumask, count, total, true, concurrent); eno != 0 {
-						return eno
-					} else {
-						atomic.AddUint32(&countDir, 1)
-					}
-				}
-			} else {
-				var dstIno2 Ino
-				if eno := m.cloneEntry(ctx, entry.Inode, entry.Attr.Typ, *dstIno, string(entry.Name), &dstIno2, cmode, cumask, count, total, true, concurrent); eno != 0 {
-					return eno
-				}
+			m.parseAttr(&pn, &pattr)
+			if pattr.Typ != TypeDirectory {
+				return syscall.ENOTDIR
 			}
-			if ctx.Canceled() {
-				return syscall.EINTR
+			if (pattr.Flags & FlagImmutable) != 0 {
+				return syscall.EPERM
```

Comment:
check if the new edge exists, just like that in Redis

Reply:
the mustInsert will do that